[SPARK-36693][REPL] Implement spark-shell idle timeouts

gyogal opened the pull request:
### What changes were proposed in this pull request?

Add a shell timeout feature via a configuration value that closes spark-shell and the context unless new lines are entered for a set amount of time.

### Why are the changes needed?

Customers have been asking for this feature because they feel like implementing session timeouts with dynamic allocation is not satisfactory and they would like to close any idle spark-shells automatically.

### Does this PR introduce _any_ user-facing change?

It adds a new config value `spark.repl.inactivityTimeout` that enables a timer for user input. Once this expires, an error is logged and the shell is closed.

### How was this patch tested?

The change was tested manually and two new unit tests have also been added.

gyogal committed:
[SPARK-36693][REPL] Implement spark-shell idle timeouts

AmplabJenkins commented:
Can one of the admins verify this patch?

attilapiros commented:
ok to test

SparkQA commented:
**[Test build #143079 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143079/testReport)** for PR 33936 at commit [`d19003c`](https://github.com/apache/spark/commit/d19003c132eb666009976d6a157277a24e21150f).

attilapiros commented:
Apache Spark leverages the resources of GitHub Actions in your forked repository to test your PR. Please enable it, see also https://github.com/apache/spark/pull/33936/checks?check_run_id=3542492885

SparkQA commented:
**[Test build #143079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143079/testReport)** for PR 33936 at commit [`d19003c`](https://github.com/apache/spark/commit/d19003c132eb666009976d6a157277a24e21150f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

attilapiros commented:
As I see this is your first PR for Apache Spark welcome here and thank you for your contribution!

attilapiros commented on repl/src/main/scala/org/apache/spark/repl/InactivityTimeout.scala:
```diff
@@ -0,0 +1,55 @@
+  def startInactivityTimer(): Unit = {
+    if (inactivityTimeoutMs > 0) {
+      inactivityTimer.foreach(_.cancel())
+      inactivityTimer = Option(new Timer(true))
```
We can name this Timer so:

```suggestion
      inactivityTimer = Some(new Timer("SparkShellInactivityTimer", isDaemon = true))
```

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143079/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47581/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47581/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/47581/

attilapiros commented on repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala:
```diff
@@ -55,7 +55,14 @@ class ReplSuite extends SparkFunSuite with BeforeAndAfterAll {
-    val in = new BufferedReader(new StringReader(input + "\n"))
+    class SlowBufferedReader(in: Reader, readLineDelay: Long) extends BufferedReader(in) {
+      override def readLine(): String = {
+        if (readLineDelay > 0) Thread.sleep(readLineDelay)
```
"Put curly braces even around one-line conditional or loop statements" 
As suggested in https://github.com/databricks/scala-style-guide#curly-braces

attilapiros commented on repl/src/main/scala/org/apache/spark/repl/InactivityTimeout.scala:
```diff
@@ -0,0 +1,55 @@
+
+package org.apache.spark.repl
+
+import java.util.{Timer, TimerTask}
```
Nit: "Avoid using wildcard imports, unless you are importing more than 6 entities, or implicit methods"

As suggested at https://github.com/databricks/scala-style-guide#imports

attilapiros replied:
As I see I have linked the wrong rule. You can leave it as it is.

attilapiros commented on repl/src/main/scala/org/apache/spark/repl/InactivityTimeout.scala:
```diff
@@ -0,0 +1,55 @@
+import java.util.{Timer, TimerTask}
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.repl.Main.{conf, sparkContext}
```
Please import just `Main` and use `Main.conf` and `Main.sparkContext` where it is needed.

attilapiros commented on repl/src/main/scala-2.13/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.repl
 
+import org.apache.spark.repl.InactivityTimeout.{startInactivityTimer, stopInactivityTimer}
```
Please import just `InactivityTimeout` and use `InactivityTimeout.startInactivityTimer` and `InactivityTimeout.stopInactivityTimer` where it is needed.

attilapiros commented on repl/src/main/scala-2.12/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -34,6 +34,8 @@ import scala.tools.nsc.interpreter.StdReplTags.tagOfIMain
 import scala.tools.nsc.util.stringFromStream
 import scala.util.Properties.{javaVersion, javaVmName, versionString}
 
+import org.apache.spark.repl.InactivityTimeout.{startInactivityTimer, stopInactivityTimer}
```
Please import just InactivityTimeout and use InactivityTimeout.startInactivityTimer and InactivityTimeout.stopInactivityTimer where it is needed.

attilapiros commented:
cc @srowen

srowen reviewed (commented):
I'm afraid this might kill the shell if a command is running for a long time? it triggers on commands.

attilapiros commented:
> I'm afraid this might kill the shell if a command is running for a long time? it triggers on commands.

@srowen 

I do not think as before the command is started the timer is stopped:
```
    stopInactivityTimer()
    super.command(line)
    startInactivityTimer()
```

srowen commented:
Ah right OK. I'm still kind of uneasy that there is some case this will kill the shell when it isn't desired, but I guess it's disabled by default.

is this not better managed by killing idle _clusters_ ? Elaborate the use case here.

AmplabJenkins commented:
Can one of the admins verify this patch?

gyogal committed:
[SPARK-36693][REPL] Spark Inactivity Timeout review fixes

SparkQA commented:
**[Test build #143088 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143088/testReport)** for PR 33936 at commit [`40c9a23`](https://github.com/apache/spark/commit/40c9a23c2f779728b1eed42d2e6e20fdba8be20c).

gyogal commented:
Thank you for the review comments! I have pushed the changes for some of them, I am unsure how to tackle the import line in SparkILoop.scala, because I would not want to import with a wildcard (as there is a rule against that). I can split the import into two lines if that works better.

The use case for customers is specifically to free up the driver resources (even though they're not managed by yarn, those are still resources being used on the gateway machines). The feature is disabled by default and the timer is stopped while commands are running in the shell and restarted when the prompt is returned.

SparkQA commented:
**[Test build #143088 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143088/testReport)** for PR 33936 at commit [`40c9a23`](https://github.com/apache/spark/commit/40c9a23c2f779728b1eed42d2e6e20fdba8be20c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143088/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47590/

srowen commented:
PS have you tested the 2.13 shell / build?

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47590/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/47590/

AmplabJenkins commented:
Can one of the admins verify this patch?

gyogal commented:
Unfortunately it seems one of the two unit tests I added fails with Scala 2.13. This is because after SPARK-30090, the prompt is not yet available after the welcome message, so it is too early to start the timer then. In the 2.11/2.12 REPL it is possible to start running commands once the message is printed but the 2.13 code initializes the Spark session afterwards which takes a couple of seconds and this change in behavior is not what the unit test expects. 

I will look into this and fix this in an update to this PR.

squito commented on repl/src/test/scala/org/apache/spark/repl/ReplSuite.scala:
```diff
@@ -382,4 +391,41 @@ class ReplSuite extends SparkFunSuite with BeforeAndAfterAll {
+
+  test("inactivity timeout is triggered") {
+    InactivityTimeout.inactivityTimeoutMs = 500
+    InactivityTimeout.isTest = true
```
rather than making these `var`s and setting them like this, can you make them confs which you set inside the test?  Then they can be `val`s inside the main code.

If you do set them like this, then they have to be reset back to the original value in a `finally` or `afterEach` etc.

gyogal replied:
Thanks for the review! I have updated the PR and reworked this part to get rid of the vars. Could you please let me know if it looks better now?

gengliangwang commented on docs/configuration.md:
```diff
@@ -2985,6 +2985,21 @@ External users can query the static sql config values via `SparkSession.conf` or
+  <td>0s</td>
+  <td>
+    Idle timeout for user input in spark-shell. The shell and the SparkContext will be closed if no new lines are
+    entered for the specified amount of time. The default value (0s) disables the feature.
```
Let's mention the time unit is milliseconds here. 
The default value is 0ms (though the value is the same as 0s)

gyogal replied:
Thanks for your review! I have pushed a change where I added that the value is milliseconds based. Please let me know if it looks good to you know or if the description should be further updated.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/47827/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/47827/

gyogal committed:
[SPARK-36693][REPL] Addressing review comments and fixing the feature under Scala 2.13

 - The vars have been removed from InquiryTimeout and it has been
   transformed into a class instead of an object.
 - The same code now works for both the Scala 2.12 and 2.13 REPLs and
   unit tests are passing.
 - The documentation has been updated to mention milliseconds as the
   time unit.

SparkQA commented:
**[Test build #143557 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143557/testReport)** for PR 33936 at commit [`3dc636b`](https://github.com/apache/spark/commit/3dc636b45c7b36b05ef2213df991298f13b4910a).

SparkQA commented:
**[Test build #143557 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143557/testReport)** for PR 33936 at commit [`3dc636b`](https://github.com/apache/spark/commit/3dc636b45c7b36b05ef2213df991298f13b4910a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class InactivityTimeout(inactivityTimeoutMs: Long) extends Logging `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143557/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48066/

gyogal committed:
[SPARK-36693][REPL] Increase timeout to make sure the unit test is not flaky

Slight update to the existing test

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48066/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48066/

SparkQA commented:
**[Test build #143565 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143565/testReport)** for PR 33936 at commit [`46eadec`](https://github.com/apache/spark/commit/46eadecbc716a599a65a5093869284467deec528).

SparkQA commented:
**[Test build #143565 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143565/testReport)** for PR 33936 at commit [`46eadec`](https://github.com/apache/spark/commit/46eadecbc716a599a65a5093869284467deec528).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143565/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48075/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48075/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48075/

attilapiros commented on repl/src/main/scala-2.12/org/apache/spark/repl/SparkILoop.scala:
```diff
@@ -105,6 +108,13 @@ class SparkILoop(in0: Option[BufferedReader], out: JPrintWriter)
 
+  override def processLine(line: String): Boolean = {
+    inactivityTimeout.stopInactivityTimer()
+    val result = super.processLine(line)
```
What happens when the line is evaluated and it throws an exception? 
Do the underlying `processLine` guarantees that exceptions are not propagated into the caller? Because if there is no such guarantees then we should wrap this into a `try {..} finally {...}` and stop the timer in the `finally`.

gyogal replied:
Thanks for your feedback! I have now updated the PR with the suggested change.

gyogal committed:
[SPARK-36693][REPL] Use a finally block to make sure the timer is started even if an interpreter exception occurs

SparkQA commented:
**[Test build #143680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143680/testReport)** for PR 33936 at commit [`95f8588`](https://github.com/apache/spark/commit/95f8588039f86e520c406504882e153133f0fb82).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143680/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48194/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48194/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48194/

gyogal committed:
[SPARK-36693][REPL] Add finally block in TimerTask to make sure shell exits even if context close throws an exception

SparkQA commented:
**[Test build #143708 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143708/testReport)** for PR 33936 at commit [`c286fb7`](https://github.com/apache/spark/commit/c286fb7ead5eeaaea9b3443300eee6e27be701c9).

SparkQA commented:
**[Test build #143708 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/143708/testReport)** for PR 33936 at commit [`c286fb7`](https://github.com/apache/spark/commit/c286fb7ead5eeaaea9b3443300eee6e27be701c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/143708/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48221/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/48221/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/48221/

attilapiros commented:
I was testing this PR and there is problem. When the shell is just started and no expressions are evaluated but only empty lines are entered the inactivity timer are not stopped:

```
$ ./bin/spark-shell --conf "spark.repl.inactivityTimeout=3s"                                                                                                                                                       [INSERT]
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
21/10/01 19:15:21 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://192.168.0.17:4040
Spark context available as 'sc' (master = local[*], app id = local-1633108523276).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.3.0-SNAPSHOT
      /_/

Using Scala version 2.12.14 (OpenJDK 64-Bit Server VM, Java 1.8.0_282)
Type in expressions to have them evaluated.
Type :help for more information.

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala>

scala> 21/10/01 19:15:28 ERROR InactivityTimeout: Inactivity timeout of 3000 ms reached - closing shell
 
```

attilapiros commented:
Thanks @gyogal working on this! Now I have closed this and when we have a good solution we can open a new PR for this.
